# Laboratory Task 2 (01_Understanding Deep Learning)

<div class="alert alert-block alert-success" style="font-family: Times New Roman">
    <h4><strong>Laboratory Task 2</strong></h4>

<p style="font-family:Times New Roman; text-align:justify; font-size:15px">
    <b>Instruction:</b> Perform a single forward pass and compute for the error.
</p>

$$x = \begin{bmatrix} 1 \\ 0 \\ 1 \end{bmatrix}$$

$$y = \begin{bmatrix} 1 \end{bmatrix}$$

$$f=max(0, Z_n)$$

$$\text{hidden unit weights} = 
\begin{bmatrix}
w_{11} = 0.2 && w_{12} = -0.3 \\
w_{13} = 0.4 && w_{14} = 0.1 \\
w_{15} = -0.5 && w_{16} = 0.2
\end{bmatrix}$$

$$\text{output unit weights} = 
\begin{bmatrix}
w_{21} = -0.3 \\
w_{22} = -0.2
\end{bmatrix}$$

$$\theta = 
\begin{bmatrix}
\theta_{1} = -0.4 \\
\theta_{2} = 0.2 \\
\theta_{3} = 0.1
\end{bmatrix}$$
</div>

## 1. Setup and Input Data

This section initializes the environment and the data. We import NumPy to handle the matrix math required for neural networks. We then define our input vector x (containing three features) and our target label y (the correct answer the network is trying to guess).

In [1]:
import numpy as np

# Inputs and Targets
x = np.array([1, 0, 1])
y = np.array([1])

## 2. Initializing Weights and Biases

Here, we construct the architecture of the network by defining its parameters. w_hidden is a 3x2 matrix because it maps our 3 inputs to 2 hidden nodes. w_output maps those 2 hidden nodes to a single output prediction. The theta variables are the biases, which act as constants added to each node to shift the activation function left or right.

In [2]:
# Weights and Biases
w_hidden = np.array([[ 0.2, -0.3],
                     [ 0.4,  0.1],
                     [-0.5,  0.2]])

w_output = np.array([-0.3, -0.2])

theta_hidden = np.array([-0.4, 0.2])
theta_output = np.array([0.1])

## 3. Computing the Hidden Layer (Forward Pass)

This is the first computational step of the forward pass. np.dot(x, w_hidden) calculates the weighted sum of the inputs, and theta_hidden is added to complete the linear transformation ($Z$). We then apply the ReLU (Rectified Linear Unit) non-linear activation function using np.maximum(0, Z_hidden). This converts any negative pre-activation values to zero, determining which signals pass through to the next layer.

In [3]:
# Hidden Layer Forward Pass
Z_hidden = np.dot(x, w_hidden) + theta_hidden
H = np.maximum(0, Z_hidden)  # ReLU activation

## 4. Computing the Output Layer (Forward Pass)

The network uses the surviving signals from the hidden layer (H) as the inputs for the final layer. It computes the dot product of H and the output weights, adds the output bias, and applies the ReLU activation function one last time to generate the network's final prediction, y_hat.

In [4]:
# Output Layer Forward Pass
Z_output = np.dot(H, w_output) + theta_output
y_hat = np.maximum(0, Z_output)  # ReLU activation

## 5. Error Calculation and Output

To determine how well the network performed, we calculate the raw difference (delta) between the prediction (y_hat) and the actual target (y). In a full training loop, this negative or positive delta would be passed backward through the network to adjust the weights. Finally, the script prints the intermediate activations and the final error.

In [5]:
# Error Calculation
delta = y_hat - y

print(f"Hidden Layer Activations (H): {H}")
print(f"Predicted Output (y_hat): {y_hat}")
print(f"Error (delta): {delta}")

Hidden Layer Activations (H): [0.  0.1]
Predicted Output (y_hat): [0.08]
Error (delta): [-0.92]


**Hidden Layer Activation Dynamics** (H = [0.0, 0.1])The output reveals how the Rectified Linear Unit (ReLU) activation function controls the network's signal flow. The pre-activation sum for the first hidden node results in a negative value (-0.7). Because ReLU is defined as $max(0, Z_n)$, it completely deactivates this neuron, resulting in a 0.0 output. The second hidden node yields a positive pre-activation sum (0.1), allowing the signal to pass through unchanged. This demonstrates ReLU's role in introducing non-linearity by dropping negative signals.

**Output Layer Prediction** (y_hat = [0.08])The final prediction heavily depends on the surviving signals from the hidden layer. Since the first hidden node outputted a 0, its corresponding weight (-0.3) is completely neutralized. The output calculation essentially becomes a combination of the second hidden node's signal (0.1) multiplied by its weight (-0.2), added to the output bias (0.1). This results in a final network prediction of 0.08.

**Error Differential** (delta = [-0.92])The error calculation ($\delta = \hat{y} - y$) represents the raw mathematical difference between the network's prediction (0.08) and the actual target (1.0). The resulting delta of -0.92 indicates a significant underestimation by the network. Because the weights and biases were arbitrarily initialized, this high error rate is expected. In a complete training loop, this negative delta serves as the starting gradient for backpropagation, signaling the network to increase the weights and biases in the preceding layers to drive the next prediction closer to 1.0.